truffle/src/main/java/org/jruby/truffle/core/hash/EnsureSymbolKeysNode.java

```diff
@@ -0,0 +1,31 @@
+package org.jruby.truffle.core.hash;
+
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.object.DynamicObject;
+import com.oracle.truffle.api.source.SourceSection;
+import org.jruby.truffle.RubyContext;
+import org.jruby.truffle.language.RubyGuards;
+import org.jruby.truffle.language.RubyNode;
+import org.jruby.truffle.language.control.RaiseException;
+
+public class EnsureSymbolKeysNode extends RubyNode {
+
+    @Child private RubyNode child;
+
+    public EnsureSymbolKeysNode(RubyContext context, SourceSection sourceSection, RubyNode child) {
+        super(context, sourceSection);
+        this.child = child;
+    }
+
+    @Override
+    public Object execute(VirtualFrame frame) {
+        final Object hash = child.execute(frame);
+        assert RubyGuards.isRubyHash(hash);
+        for (KeyValue keyValue : HashOperations.iterableKeyValues((DynamicObject) hash)) {
+            if (!RubyGuards.isRubySymbol(keyValue.getKey())) {
+                throw new RaiseException(coreExceptions().typeErrorWrongArgumentType(keyValue.getKey(), "Symbol", this));
```

Comment:
You want a BranchProfile here before the `throw` so that part is not compiled until we actually meet a non-Symbol.
The BranchProfile should be a final field of the node.